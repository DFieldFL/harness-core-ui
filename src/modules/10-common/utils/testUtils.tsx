import React from 'react'
import { UseGetProps, UseGetReturn, RestfulProvider } from 'restful-react'
import { compile } from 'path-to-regexp'
import { createMemoryHistory } from 'history'
import { Router, Route, Switch, useLocation } from 'react-router-dom'
import { ModalProvider } from '@wings-software/uikit'
import qs from 'qs'

import strings from 'strings/strings.en.yaml'
import { AppStoreProvider, useAppStoreWriter } from 'framework/hooks/useAppStore'
import type { AppStore } from 'framework/types/AppStore'
import { AUTH_ROUTE_PATH_PREFIX, StringsContext } from 'framework/exports'

export type UseGetMockData<TData, TError = undefined, TQueryParams = undefined, TPathParams = undefined> = Required<
  UseGetProps<TData, TError, TQueryParams, TPathParams>
>['mock']

export interface UseMutateMockData<TData, TRequestBody = unknown> {
  loading?: boolean
  mutate?: (data?: TRequestBody) => Promise<TData>
}

export type UseGetReturnData<TData, TError = undefined, TQueryParams = undefined, TPathParams = undefined> = Omit<
  UseGetReturn<TData, TError, TQueryParams, TPathParams>,
  'absolutePath' | 'cancel' | 'response'
>

export const findDialogContainer = (): Element | null => document.querySelector('.bp3-dialog')
export const findPopoverContainer = (): Element | null => document.querySelector('.bp3-popover-content')

interface TestWrapperProps {
  path?: string
  pathParams?: Record<string, string | number>
  queryParams?: Record<string, any>
  defaultAppStoreValues?: Partial<AppStore>
}

export const prependAccountPath = (path: string): string => AUTH_ROUTE_PATH_PREFIX + path

const AppStoreInitializer: React.FC<{ defaultAppStoreValues: TestWrapperProps['defaultAppStoreValues'] }> = ({
  children,
  defaultAppStoreValues
}) => {
  const updateAppStore = useAppStoreWriter()
  updateAppStore(defaultAppStoreValues || {})
  return <> {children} </>
}

export const NotFound = (): JSX.Element => {
  const location = useLocation()
  return (
    <div>
      <h1>Not Found</h1>
      <div data-testid="location">{`${location.pathname}${
        location.search ? `?${location.search.replace(/^\?/g, '')}` : ''
      }`}</div>
    </div>
  )
}

export const TestWrapper: React.FC<TestWrapperProps> = props => {
  const { path = '/', pathParams = {}, defaultAppStoreValues, queryParams = {} } = props

  const search = qs.stringify(queryParams, { addQueryPrefix: true })
  const routePath = compile(path)(pathParams) + search

  const history = createMemoryHistory({ initialEntries: [routePath] })

  return (
    <AppStoreProvider>
      <StringsContext.Provider value={strings}>
        <AppStoreInitializer defaultAppStoreValues={defaultAppStoreValues}>
          <Router history={history}>
            <ModalProvider>
              <RestfulProvider base="/">
                <Switch>
                  <Route exact path={path}>
                    {props.children}
                  </Route>
                  <Route>
                    <NotFound />
                  </Route>
                </Switch>
              </RestfulProvider>
            </ModalProvider>
          </Router>
        </AppStoreInitializer>
      </StringsContext.Provider>
    </AppStoreProvider>
  )
}
