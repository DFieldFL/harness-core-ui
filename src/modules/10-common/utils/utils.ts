import type { IconName } from '@wings-software/uicore'
import { Scope } from '@common/interfaces/SecretsInterface'
import { ModuleName } from 'framework/types/ModuleName'

interface SetPageNumberProps {
  setPage: (value: React.SetStateAction<number>) => void
  pageItemsCount?: number
  page: number
}

export const getModuleIcon = (module: ModuleName): IconName => {
  switch (module) {
    case ModuleName.CD:
      return 'cd-main'
    case ModuleName.CV:
      return 'cv-main'
    case ModuleName.CI:
      return 'ci-main'
    case ModuleName.CE:
      return 'ce-main'
    case ModuleName.CF:
      return 'cf-main'
  }
  return 'nav-project'
}

export const getReference = (scope?: Scope, identifier?: string): string | undefined => {
  switch (scope) {
    case Scope.PROJECT:
      return identifier
    case Scope.ORG:
      return `org.${identifier}`
    case Scope.ACCOUNT:
      return `account.${identifier}`
  }
}

export const setPageNumber = ({ setPage, pageItemsCount, page }: SetPageNumberProps): void => {
  if (pageItemsCount === 0 && page > 0) {
    setPage(page - 1)
  }
}
