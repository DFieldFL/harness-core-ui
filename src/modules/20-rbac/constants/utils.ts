export enum RbacResourceGroupTypes {
  DYNAMIC_RESOURCE_SELECTOR = 'DynamicResourceSelector'
}
