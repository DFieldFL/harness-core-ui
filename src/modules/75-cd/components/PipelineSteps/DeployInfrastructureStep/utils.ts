/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { FormikErrors, yupToFormErrors } from 'formik'
import { defaultTo, get, isEmpty } from 'lodash-es'
import * as Yup from 'yup'

import {
  getMultiTypeFromValue,
  MultiTypeInputType,
  RUNTIME_INPUT_VALUE,
  SelectOption,
  MultiSelectWithSubmenuOption
} from '@harness/uicore'
import type { EnvironmentResponseDTO } from 'services/cd-ng'

import { getDurationValidationSchema } from '@common/components/MultiTypeDuration/MultiTypeDuration'

import type { DeployStageConfig } from '@pipeline/utils/DeployStageInterface'
import { StepViewType, ValidateInputSetProps } from '@pipeline/components/AbstractSteps/Step'

export interface DeployInfrastructureProps {
  initialValues: DeployStageConfig
  onUpdate?: (data: DeployStageConfig) => void
  readonly: boolean
  allowableTypes: MultiTypeInputType[]
  stepViewType?: StepViewType
  inputSetData?: {
    template?: DeployStageConfig
    path?: string
    readonly?: boolean
  }
}

export const ALL_SELECTED = 'All'

export function isEditEnvironment(data?: EnvironmentResponseDTO): boolean {
  if (!isEmpty(data?.identifier)) {
    return true
  }
  return false
}

export function isEditEnvironmentOrEnvGroup(data?: EnvironmentResponseDTO): boolean {
  if (!isEmpty(data?.identifier)) {
    return true
  }
  return false
}

export function isEditInfrastructure(data?: string): boolean {
  if (!isEmpty(data)) {
    return true
  }
  return false
}

export function validateStepForm({
  data,
  template,
  getString,
  viewType
}: ValidateInputSetProps<any>): FormikErrors<any> {
  const isRequired = viewType === StepViewType.DeploymentForm || viewType === StepViewType.TriggerForm
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const errors = {} as any
  // istanbul ignore next
  // istanbul ignore else
  if (getMultiTypeFromValue(template?.timeout) === MultiTypeInputType.RUNTIME) {
    // istanbul ignore next
    let timeoutSchema = getDurationValidationSchema({ minimum: '10s' })
    // istanbul ignore next
    if (isRequired) {
      // istanbul ignore next
      timeoutSchema = timeoutSchema.required(getString?.('validation.timeout10SecMinimum'))
    }
    const timeout = Yup.object().shape({
      timeout: timeoutSchema
    })

    try {
      timeout.validateSync(data)
    } catch (e) {
      /* istanbul ignore else */
      if (e instanceof Yup.ValidationError) {
        const err = yupToFormErrors(e)

        Object.assign(errors, err)
      }
    }
  }
  return errors
}

export function processNonGitOpsInitialValues(initialValues: DeployStageConfig) {
  return {
    environment: {
      environmentRef: defaultTo(initialValues.environment?.environmentRef, ''),
      deployToAll: defaultTo(initialValues.environment?.deployToAll, false)
    },
    infrastructureRef: (initialValues.environment?.infrastructureDefinitions?.[0].ref ||
      initialValues.environment?.infrastructureDefinitions ||
      '') as string
  }
}

export function processNonGitOpsFormValues(data: DeployStageConfig) {
  return {
    environment: {
      environmentRef: data.environment?.environmentRef,
      deployToAll: data.environment?.environmentRef === RUNTIME_INPUT_VALUE ? true : false,
      ...(data.environment?.environmentRef &&
        data.environment?.environmentRef !== RUNTIME_INPUT_VALUE &&
        data.infrastructureRef && {
          infrastructureDefinitions:
            data.infrastructureRef === RUNTIME_INPUT_VALUE
              ? RUNTIME_INPUT_VALUE
              : [
                  {
                    ref: data.infrastructureRef
                  }
                ]
        })
    }
  }
}

export function processGitOpsEnvironmentInitialValues(initialValues: DeployStageConfig) {
  return {
    isEnvGroup: false,
    environmentOrEnvGroupRef: defaultTo(initialValues.environment?.environmentRef, ''),
    environmentOrEnvGroupAsRuntime: 'Environment',
    deployToAll: defaultTo(initialValues.environment?.deployToAll, false),
    clusterRef:
      getMultiTypeFromValue(initialValues.environment?.gitOpsClusters as any) === MultiTypeInputType.RUNTIME
        ? RUNTIME_INPUT_VALUE
        : initialValues.environment?.deployToAll
        ? [
            {
              label: 'All Clusters in Environment',
              value: ALL_SELECTED
            }
          ]
        : defaultTo(initialValues.environment?.gitOpsClusters, [])?.map(cluster => {
            return {
              label: cluster.ref,
              value: cluster.ref
            }
          })
  }
}

export function processGitOpsEnvironmentFormValues(data: DeployStageConfig) {
  const allClustersSelected = (data.clusterRef as SelectOption[])?.[0].value === ALL_SELECTED

  return {
    environment: {
      environmentRef:
        data.environmentOrEnvGroupRef === RUNTIME_INPUT_VALUE
          ? RUNTIME_INPUT_VALUE
          : defaultTo((data.environmentOrEnvGroupRef as SelectOption)?.value, ''),
      deployToAll: data.environmentOrEnvGroupRef === RUNTIME_INPUT_VALUE ? true : allClustersSelected,
      ...(data.environmentOrEnvGroupRef &&
        data.environmentOrEnvGroupRef !== RUNTIME_INPUT_VALUE &&
        !allClustersSelected && {
          gitOpsClusters:
            data.clusterRef === RUNTIME_INPUT_VALUE
              ? RUNTIME_INPUT_VALUE
              : (data.clusterRef as SelectOption[])?.map(cluster => ({ ref: cluster.value }))
        })
    }
  }
}

export function processGitOpsEnvGroupInitialValues(initialValues: DeployStageConfig) {
  return {
    isEnvGroup: true,
    environmentOrEnvGroupRef: defaultTo(initialValues.environmentGroup?.envGroupRef, ''),
    environmentOrEnvGroupAsRuntime: 'Environment Group',
    environmentsInEnvGroup:
      getMultiTypeFromValue(initialValues.environmentGroup?.envGroupRef) === MultiTypeInputType.RUNTIME
        ? RUNTIME_INPUT_VALUE
        : initialValues.environmentGroup?.deployToAll
        ? [
            {
              name: ALL_SELECTED
            }
          ]
        : defaultTo(initialValues.environmentGroup?.envGroupConfig, [])?.map(environment => {
            return {
              name: environment.environmentRef,
              deployToAll: environment.deployToAll,
              ...(!environment.deployToAll && { clusters: defaultTo(environment.gitOpsClusters, []) })
            }
          }),
    deployToAll: defaultTo(initialValues.environmentGroup?.deployToAll, false),
    clusterRef: initialValues.environmentGroup?.deployToAll
      ? RUNTIME_INPUT_VALUE
      : initialValues.environmentGroup?.envGroupConfig?.reduce((prev, environment) => {
          return [
            ...prev,
            ...(environment.deployToAll
              ? [
                  {
                    label: `All`,
                    value: 'All',
                    parentLabel: environment.environmentRef,
                    parentValue: environment.environmentRef
                  }
                ]
              : defaultTo(
                  environment?.gitOpsClusters?.map(cluster => ({
                    label: cluster.ref,
                    value: cluster.ref,
                    parentLabel: environment.environmentRef,
                    parentValue: environment.environmentRef
                  })),
                  []
                ))
          ]
        }, [] as MultiSelectWithSubmenuOption[])
  }
}

export function processGitOpsEnvGroupFormValues(data: DeployStageConfig) {
  const environmentClusterMap: any = {}
  ;(data.clusterRef as MultiSelectWithSubmenuOption[])?.forEach(cluster => {
    if (cluster.value !== 'All') {
      try {
        environmentClusterMap[get(cluster, 'parentValue', '')].push(cluster.value)
      } catch (e: any) {
        environmentClusterMap[get(cluster, 'parentValue', '')] = [cluster.value]
      }
    }
  })

  const allEnvironmentsSelected = (data.environmentInEnvGroupRef as SelectOption[])?.[0].value === ALL_SELECTED

  return {
    environmentGroup: {
      envGroupRef:
        data.environmentOrEnvGroupRef === RUNTIME_INPUT_VALUE
          ? RUNTIME_INPUT_VALUE
          : defaultTo((data.environmentOrEnvGroupRef as SelectOption)?.value, ''),
      deployToAll: data.environmentOrEnvGroupRef === RUNTIME_INPUT_VALUE || allEnvironmentsSelected,
      ...(!(data.environmentOrEnvGroupRef === RUNTIME_INPUT_VALUE || allEnvironmentsSelected) && {
        envGroupConfig:
          data.environmentInEnvGroupRef === RUNTIME_INPUT_VALUE
            ? RUNTIME_INPUT_VALUE
            : (data?.environmentInEnvGroupRef as SelectOption[])?.map(environmentInEnvGroup => {
                const areClustersSelected = Boolean(environmentClusterMap[environmentInEnvGroup.value])
                return {
                  environmentRef: environmentInEnvGroup?.value,
                  deployToAll: !areClustersSelected,
                  ...(areClustersSelected && {
                    gitOpsClusters:
                      data.clusterRef === RUNTIME_INPUT_VALUE
                        ? RUNTIME_INPUT_VALUE
                        : environmentClusterMap[environmentInEnvGroup.value]?.map((cluster: any) => ({ ref: cluster }))
                  })
                }
              })
      })
    }
  }
}
