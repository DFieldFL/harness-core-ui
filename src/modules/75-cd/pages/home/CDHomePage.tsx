import React from 'react'
import { HomePageTemplate } from '@common/components/HomePageTemplate/HomePageTemplate'
import { useStrings } from 'framework/strings'
import { ModuleName } from 'framework/types/ModuleName'
import bgImageURL from './images/homeIllustration.svg'

export const CDHomePage: React.FC = () => {
  const { getString } = useStrings()

  const trialBannerProps = {
    module: ModuleName.CD
  }

  return (
    <HomePageTemplate
      title={getString('cd.continuous')}
      bgImageUrl={bgImageURL}
      subTitle={getString('cd.dashboard.subHeading')}
      documentText={getString('cd.learnMore')}
      trialBannerProps={trialBannerProps}
    />
  )
}

export default CDHomePage
