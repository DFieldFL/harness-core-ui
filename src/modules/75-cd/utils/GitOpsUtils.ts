import type { GitOpsInfoDTO, ManagedArgoGitOpsInfoDTO, ConnectedArgoGitOpsInfoDTO } from 'services/cd-ng'

export type GitOpsProviderType = Required<GitOpsInfoDTO>['type']

export const GitOpsProviderTypeEnum: Record<GitOpsProviderType, GitOpsProviderType> = {
  CONNECTED_ARGO_PROVIDER: 'CONNECTED_ARGO_PROVIDER',
  MANAGED_ARGO_PROVIDER: 'MANAGED_ARGO_PROVIDER'
}

export function isConnectedGitOpsProvider(obj?: GitOpsInfoDTO | null): obj is ConnectedArgoGitOpsInfoDTO {
  return obj?.type === GitOpsProviderTypeEnum.CONNECTED_ARGO_PROVIDER
}

export function isManagedGitOpsProvider(obj?: GitOpsInfoDTO | null): obj is ManagedArgoGitOpsInfoDTO {
  return obj?.type === GitOpsProviderTypeEnum.MANAGED_ARGO_PROVIDER
}
