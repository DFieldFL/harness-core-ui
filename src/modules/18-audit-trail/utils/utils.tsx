import React from 'react'
import { Menu } from '@blueprintjs/core'
import type { ItemRenderer } from '@blueprintjs/select'
import { Layout, Text, Avatar, Color, MultiSelectOption } from '@wings-software/uicore'
import { isEmail } from '@common/utils/Validation'

export interface UserItem extends MultiSelectOption {
  email?: string
}

export const UserItemRenderer: ItemRenderer<UserItem> = (item, { handleClick }) => (
  <Menu.Item
    key={item.value.toString()}
    text={
      <Layout.Horizontal spacing="small" flex={{ alignItems: 'center', justifyContent: 'flex-start' }}>
        <Avatar name={item.label} email={item.email || item.value.toString()} size="small" hoverCard={false} />
        <Layout.Vertical padding={{ left: 'small' }}>
          <Text color={Color.BLACK}>{item.label}</Text>
          <Text color={Color.GREY_700}>{item.email || item.value}</Text>
        </Layout.Vertical>
      </Layout.Horizontal>
    }
    onClick={handleClick}
  />
)

export const UserTagRenderer = (item: UserItem, validate = false): React.ReactNode => (
  <Layout.Horizontal key={item.value.toString()} flex spacing="small">
    <Avatar name={item.label} email={item.value.toString()} size="xsmall" hoverCard={false} />
    <Text color={validate && !isEmail(item.value.toString().toLowerCase()) ? Color.RED_500 : Color.BLACK}>
      {item.label}
    </Text>
  </Layout.Horizontal>
)
