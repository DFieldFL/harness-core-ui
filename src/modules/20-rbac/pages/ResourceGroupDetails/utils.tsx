import { RbacResourceGroupTypes } from '@rbac/constants/utils'
import type { ResourceType } from '@rbac/interfaces/ResourceType'
import type { ResourceSelector } from 'services/cd-ng'

export const getSelectedResourcesMap = (
  resourceSelectorList?: ResourceSelector[]
): Map<ResourceType, string[] | string> => {
  const map: Map<ResourceType, string[] | string> = new Map()

  resourceSelectorList?.map(resourceSelector => {
    if (resourceSelector.type === RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
      map.set(resourceSelector.resourceType, RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
    else map.set(resourceSelector.resourceType, resourceSelector.identifiers)
  })

  return map
}

export const getResourceSelectorsfromMap = (map: Map<ResourceType, string[] | string>): ResourceSelector[] => {
  const resourceSelectors: ResourceSelector[] = []
  map.forEach((value, key) => {
    if (value === RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
      resourceSelectors.push({
        type: RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR,
        resourceType: key
      })
    else {
      resourceSelectors.push({
        type: RbacResourceGroupTypes.STATIC_RESOURCE_SELECTOR,
        resourceType: key,
        identifiers: value
      })
    }
  })
  return resourceSelectors
}
