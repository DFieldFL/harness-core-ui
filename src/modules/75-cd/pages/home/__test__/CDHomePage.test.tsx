import React from 'react'
import { render } from '@testing-library/react'
import { TestWrapper } from '@common/utils/testUtils'
import CDHomePage from '../CDHomePage'

describe('CDHomePage snapshot test', () => {
  test('should render properly', async () => {
    const { container } = render(
      <TestWrapper pathParams={{ orgIdentifier: 'dummy' }}>
        <CDHomePage />
      </TestWrapper>
    )
    expect(container).toMatchSnapshot()
  })
})
