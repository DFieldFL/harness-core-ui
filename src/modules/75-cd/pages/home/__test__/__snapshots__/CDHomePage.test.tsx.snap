// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CDHomePage snapshot test should render properly 1`] = `
<div>
  <div
    class="StyledProps--font StyledProps--main"
    style="margin: 80px; background: transparent url(test-file-stub) no-repeat; position: relative; background-size: contain; background-position-y: center; height: calc(100% - 160px);"
  >
    <div
      class="Layout--vertical Layout--layout-spacing-medium StyledProps--main"
    >
      <h1
        class="StyledProps--font StyledProps--font-h1 StyledProps--main StyledProps--font-weight-bold StyledProps--color"
      >
        cd.continuous
      </h1>
      <p
        class="StyledProps--font StyledProps--main StyledProps--color StyledProps--font-size-normal"
        style="line-height: 24px; width: 400px;"
      >
        cd.dashboard.subHeading
      </p>
      <a
        class="bp3-button Button--button StyledProps--font StyledProps--main StyledProps--color StyledProps--font StyledProps--font-size-normal Button--with-current-color Button--link"
        href="https://ngdocs.harness.io/"
        rel="noreferrer noopener"
        role="button"
        style="align-self: flex-start;"
        tabindex="0"
        target="_blank"
      >
        <span
          class="bp3-button-text"
        >
          cd.learnMore
        </span>
      </a>
      <div
        class="Layout--horizontal Layout--layout-spacing-large StyledProps--main StyledProps--flex"
      >
        <a
          class="createBtn"
          href="/account/undefined/projects"
        >
          createProject
        </a>
        <p
          class="StyledProps--font StyledProps--main StyledProps--font-size-medium StyledProps--color StyledProps--color-black"
        >
          orSelectExisting
        </p>
        <span
          style="flex-grow: 1;"
        />
      </div>
    </div>
  </div>
</div>
`;
